consul/acl.go

```diff
@@ -192,3 +194,168 @@ func (s *Server) useACLPolicy(id, authDC string, cached *aclCacheEntry, p *struc
+			continue
+		}
+		f.logger.Printf("[DEBUG] consul: dropping node %q from result due to ACLs", node.Node)
+		sn = append(sn[:i], sn[i+1:]...)
+		i--
+	}
+	*nodes = sn
+}
+
+// filterNodeServices is used to filter services on a given node base on ACLs.
+func (f *aclFilter) filterNodeServices(services *structs.NodeServices) {
+	for svc, _ := range services.Services {
+		if f.filterService(svc) {
+			continue
+		}
+		f.logger.Printf("[DEBUG] consul: dropping service %q from result due to ACLs", svc)
+		delete(services.Services, svc)
+	}
+}
+
+// filterCheckServiceNodes is used to filter nodes based on ACL rules.
+func (f *aclFilter) filterCheckServiceNodes(nodes *structs.CheckServiceNodes) {
+	csn := *nodes
+	for i := 0; i < len(csn); i++ {
+		node := csn[i]
+		if f.filterService(node.Service.Service) {
+			continue
+		}
+		f.logger.Printf("[DEBUG] consul: dropping node %q from result due to ACLs", node.Node.Node)
+		csn = append(csn[:i], csn[i+1:]...)
+		i--
+	}
+	*nodes = csn
+}
+
+// filterNodeDump is used to filter through all parts of a node dump and
+// remove elements the provided ACL token cannot access.
+func (f *aclFilter) filterNodeDump(dump *structs.NodeDump) {
+	nd := *dump
+	for i := 0; i < len(nd); i++ {
+		info := nd[i]
+
+		// Filter services
+		for i := 0; i < len(info.Services); i++ {
+			svc := info.Services[i].Service
+			if f.filterService(svc) {
+				continue
+			}
+			f.logger.Printf("[DEBUG] consul: dropping service %q from result due to ACLs", svc)
+			info.Services = append(info.Services[:i], info.Services[i+1:]...)
+			i--
+		}
+
+		// Filter checks
+		for i := 0; i < len(info.Checks); i++ {
+			chk := info.Checks[i]
+			if f.filterService(chk.ServiceName) {
+				continue
+			}
+			f.logger.Printf("[DEBUG] consul: dropping check %q from result due to ACLs", chk.CheckID)
+			info.Checks = append(info.Checks[:i], info.Checks[i+1:]...)
+			i--
+		}
+	}
+	*dump = nd
+}
+
+// filterACL is used to filter results from our service catalog based on the
+// rules configured for the provided token. The subject is scrubbed and
+// modified in-place, leaving only resources the token can access.
+func (s *Server) filterACL(token string, subj interface{}) error {
+	// Get the ACL from the token
+	acl, err := s.resolveToken(token)
+	if err != nil {
+		return err
+	}
+
+	// Fast path if ACLs are not enabled
+	if acl == nil {
+		return nil
+	}
+
+	// Create the filter
+	filt := newAclFilter(acl, s.logger)
+
+	switch v := subj.(type) {
+	case *structs.IndexedHealthChecks:
+		filt.filterHealthChecks(&v.HealthChecks)
+
+	case *structs.IndexedServices:
+		filt.filterServices(v.Services)
+
+	case *structs.IndexedServiceNodes:
+		filt.filterServiceNodes(&v.ServiceNodes)
+
+	case *structs.IndexedNodeServices:
+		filt.filterNodeServices(v.NodeServices)
+
+	case *structs.IndexedCheckServiceNodes:
+		filt.filterCheckServiceNodes(&v.Nodes)
+
+	case *structs.IndexedNodeDump:
+		filt.filterNodeDump(&v.Dump)
+	}
```

Comment:
This seems like the kind of thing that should freak out if a new type is encountered (we could add `case` clauses for ones that we explicitly don't care to filter).